validator/client/validator.go

```diff
@@ -369,6 +370,94 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
 	return time.Unix(int64(v.genesisTime), 0 /*ns*/).Add(secs * time.Second)
 }
 
+// CheckDoppelGanger checks if the current actively provided keys have
+// any duplicates active in the network.
+func (v *validator) CheckDoppelGanger(ctx context.Context) error {
+	if !featureconfig.Get().EnableDoppelGanger {
+		return nil
+	}
+	pubkeys, err := v.keyManager.FetchValidatingPublicKeys(ctx)
+	if err != nil {
+		return err
+	}
+	log.WithField("keys", len(pubkeys)).Info("Running doppelganger check")
+	// Exit early if no validating pub keys are found.
+	if len(pubkeys) == 0 {
+		return nil
+	}
+	req := &ethpb.DoppelGangerRequest{ValidatorRequests: []*ethpb.DoppelGangerRequest_ValidatorRequest{}}
+	for _, pkey := range pubkeys {
+		attRec, err := v.db.AttestationHistoryForPubKey(ctx, pkey)
+		if err != nil {
+			return err
+		}
+		if len(attRec) == 0 {
+			// If no history exists we simply send in a zero
+			// value for the request epoch and root.
+			req.ValidatorRequests = append(req.ValidatorRequests,
+				&ethpb.DoppelGangerRequest_ValidatorRequest{
+					PublicKey:  pkey[:],
+					Epoch:      0,
+					SignedRoot: make([]byte, 32),
+				})
+			continue
+		}
+		r := retrieveLatestRecord(attRec)
+		req.ValidatorRequests = append(req.ValidatorRequests,
+			&ethpb.DoppelGangerRequest_ValidatorRequest{
+				PublicKey:  r.PubKey[:],
+				Epoch:      r.Target,
+				SignedRoot: r.SigningRoot[:],
+			})
```

Comment:
r.PubKey is always zero bytes. It is not set here: https://github.com/prysmaticlabs/prysm/blob/ca7b3127613bf9522d1b49dce4b1142f41702efe/validator/db/kv/attester_protection.go#L116-L119

Reply:
I've fixed this PR, but now other tests are failing due to this bug. I'll submit another PR to develop to fix the underlying issue with attester_protection.go AttestationHistoryForPubKey, then we can resync this PR with develop.

Reply:
#9135 will fix this bug and round trip tests.